src/_jpegxl.c

```diff
@@ -0,0 +1,629 @@
+            }
+            return "RGBA";
+        }
+        if (bi->num_color_channels == 1) {
+            if (bi->alpha_premultiplied) {
+                return "La";
+            }
+            return "LA";
+        }
+    }
+
+    // image has no transparency
+    if (bi->num_color_channels == 3) {
+        return "RGB";
+    }
+    if (bi->num_color_channels == 1) {
+        return "L";
+    }
+
+    // could not recognize mode
+    return NULL;
+}
+
+// Decoder type
+typedef struct {
+    PyObject_HEAD JxlDecoder *decoder;
+    void *runner;
+
+    uint8_t *jxl_data;        // input jxl bitstream
+    Py_ssize_t jxl_data_len;  // length of input jxl bitstream
+
+    uint8_t *outbuf;
+    Py_ssize_t outbuf_len;
+
+    uint8_t *jxl_icc;
+    Py_ssize_t jxl_icc_len;
+    uint8_t *jxl_exif;
+    Py_ssize_t jxl_exif_len;
+    uint8_t *jxl_xmp;
+    Py_ssize_t jxl_xmp_len;
+
+    JxlDecoderStatus status;
+    JxlBasicInfo basic_info;
+    JxlPixelFormat pixel_format;
+
+    Py_ssize_t n_frames;
+
+    char *mode;
+} JpegXlDecoderObject;
+
+static PyTypeObject JpegXlDecoder_Type;
+
+void
+_jxl_decoder_dealloc(PyObject *self) {
+    JpegXlDecoderObject *decp = (JpegXlDecoderObject *)self;
+
+    if (decp->jxl_data) {
+        free(decp->jxl_data);
+        decp->jxl_data = NULL;
+        decp->jxl_data_len = 0;
+    }
+    if (decp->outbuf) {
+        free(decp->outbuf);
+        decp->outbuf = NULL;
+        decp->outbuf_len = 0;
+    }
+    if (decp->jxl_icc) {
+        free(decp->jxl_icc);
+        decp->jxl_icc = NULL;
+        decp->jxl_icc_len = 0;
+    }
+    if (decp->jxl_exif) {
+        free(decp->jxl_exif);
+        decp->jxl_exif = NULL;
+        decp->jxl_exif_len = 0;
+    }
+    if (decp->jxl_xmp) {
+        free(decp->jxl_xmp);
+        decp->jxl_xmp = NULL;
+        decp->jxl_xmp_len = 0;
+    }
+
+    if (decp->decoder) {
+        JxlDecoderDestroy(decp->decoder);
+        decp->decoder = NULL;
+    }
+
+    if (decp->runner) {
+        JxlThreadParallelRunnerDestroy(decp->runner);
+        decp->runner = NULL;
+    }
+}
+
+// sets input jxl bitstream loaded into jxl_data
+// has to be called after every rewind
+void
+_jxl_decoder_set_input(PyObject *self) {
+    JpegXlDecoderObject *decp = (JpegXlDecoderObject *)self;
+
+    decp->status =
+        JxlDecoderSetInput(decp->decoder, decp->jxl_data, decp->jxl_data_len);
+
+    // the input contains the whole jxl bitstream so it can be closed
+    JxlDecoderCloseInput(decp->decoder);
+}
+
+PyObject *
+_jxl_decoder_rewind(PyObject *self) {
+    JpegXlDecoderObject *decp = (JpegXlDecoderObject *)self;
+    JxlDecoderRewind(decp->decoder);
+    Py_RETURN_NONE;
+}
+
+bool
+_jxl_decoder_count_frames(PyObject *self) {
+    JpegXlDecoderObject *decp = (JpegXlDecoderObject *)self;
+
+    decp->n_frames = 0;
+
+    // count all JXL_DEC_NEED_IMAGE_OUT_BUFFER events
+    while (decp->status != JXL_DEC_SUCCESS) {
+        decp->status = JxlDecoderProcessInput(decp->decoder);
```

Comment:
I actually think it would probably be preferable to not call `JxlDecoderProcessInput` on open at all, to improve performance, and only set `n_frames` when the user asks for it.

https://github.com/python-pillow/Pillow/pull/7311#issuecomment-1723004045
> For me this is one of the key concepts in Pillow: you can safely open the image file without decoding the whole image (which is much more expensive operation) and get any info from the file.

That could be as simple as https://github.com/python-pillow/Pillow/blob/3c71559804e661a5f727e2007a5be51f26d9af27/src/PIL/GifImagePlugin.py#L125-L135